### ISSUE TITLE
rustdoc: reexported type aliases are not preserved in function signatures

### ISSUE THREAD

**Issue by ghost** [rust-lang/rust#15823]
The recently added `iterate` function uses the `Iterate` type alias as its result type. It is correctly displayed in http://static.rust-lang.org/doc/master/core/iter/fn.iterate.html but not in  http://static.rust-lang.org/doc/master/std/iter/fn.iterate.html.

**Comment by steveklabnik** [comment]
This file:

```
pub use std::result::Result as Lol;
```

still mis-renders:

![2015-11-07-081909_380x121_scrot](https://cloud.githubusercontent.com/assets/27786/11014100/2f66c164-8528-11e5-8afb-85caa168dd96.png)


though as @icorderi mentions, the titles are correct.

**Comment by steveklabnik** [comment]
Triage: no changes since my last comment

**Comment by GuillaumeGomez** [comment]
Seems to be working now so closing.

**Comment by svartalf** [comment]
I don't think this issue is fully fixed (correct me if I'm wrong); for example, for [`heim`](https://docs.rs/heim/0.0.10/heim/) docs.rs does not render type alias correctly, resulting in a weirdly gibberish output: https://docs.rs/heim/0.0.10/heim/cpu/struct.CpuFrequency.html#method.current

Type alias is used in the sub-crate code: https://github.com/heim-rs/heim/blob/095768ff6a580b27727b6958521392bb5f86a443/heim-cpu/src/freq.rs#L20-L22

And whole this crate is re-exported with `#[doc(inline)]` later at the facade crate: https://github.com/heim-rs/heim/blob/v0.0.10/heim/src/lib.rs#L59-L61

Docs for sub-crate are rendered almost correctly: https://docs.rs/heim-cpu/0.0.10/heim_cpu/struct.CpuFrequency.html#method.current
It does not has an active link, though, which might be part of the problem.

**Comment by svartalf** [comment]
Here is a minimally working example for this issue: https://github.com/svartalf/rust-issue-15823

**Comment by GuillaumeGomez** [comment]
It works fine for me? Version 1.43.

**Comment by svartalf** [comment]
@GuillaumeGomez thanks for getting back to this issue!

Unfortunately, I still can reproduce it with the example from my previous comment and `rust version 1.43.0-nightly (4ad624882 2020-03-03)`:
![image](https://user-images.githubusercontent.com/1279564/75890210-16ff4480-5e3f-11ea-9a10-08f95d4d712d.png).
You can check the [example repo](https://github.com/svartalf/rust-issue-15823#problem) for additional details and screenshots.

Same happens at docs.rs for `heim` crate: https://docs.rs/heim/0.1.0-alpha.1/heim/cpu/struct.CpuFrequency.html; I would expect to see [`Frequency`](https://docs.rs/heim/0.1.0-alpha.1/heim/units/type.Frequency.html) type alias instead of `Quantity<dyn Dimension<M = Z0, I = Z0, N = Z0, ..` as it is shown right now.

**Comment by alice-i-cecile** [comment]
This is still an issue for me on rust 1.60.0-nightly.

**Comment by GuillaumeGomez** [comment]
It seems to be a tricky issue to fix. I opened https://github.com/rust-lang/compiler-team/issues/504 so we'll see if it'll be able to fix this problem.

In the meantime, to add a test for this in rustdoc:

### src/test/rustdoc/alias-reexport.rs

```rust
// aux-build:alias-reexport.rs
// aux-build:alias-reexport2.rs

#![crate_name = "foo"]

extern crate alias_reexport2;

// @has 'foo/reexport/fn.foo.html'
// @has - '//*[@class="docblock item-decl"]' 'pub fn foo() -> Reexport'
// @has 'foo/reexport/fn.foo2.html'
// @has - '//*[@class="docblock item-decl"]' 'pub fn foo2() -> Result<Reexport, ()>'
// @has 'foo/reexport/type.Reexported.html'
// @has - '//*[@class="docblock item-decl"]' 'pub type Reexported'
#[doc(inline)]
pub use alias_reexport2 as reexport;
```

### src/test/rustdoc/alias-reexport2.rs

```rust
// aux-build:alias-reexport.rs

#![crate_name = "foo"]

extern crate alias_reexport;

use alias_reexport::Reexported;

// @has 'foo/fn.foo.html'
// @has - '//*[@class="docblock item-decl"]' 'pub fn foo() -> Reexported'
pub fn foo() -> Reexported { 0 }
// @has 'foo/fn.foo2.html'
// @has - '//*[@class="docblock item-decl"]' 'pub fn foo2() -> Result<Reexported, ()>'
pub fn foo2() -> Result<Reexported, ()> { Ok(0) }
```

### src/test/rustdoc/auxiliary/alias-reexport.rs

```rust
pub type Reexported = u8;
```

### src/test/rustdoc/auxiliary/alias-reexport2.rs

```rust
extern crate alias_reexport;

pub use alias_reexport::Reexported;

// @has 'foo/fn.foo.html'
// @has - '//*[@class="docblock item-decl"]' 'pub fn foo() -> Reexported'
pub fn foo() -> Reexported { 0 }
// @has 'foo/fn.foo2.html'
// @has - '//*[@class="docblock item-decl"]' 'pub fn foo2() -> Result<Reexported, ()>'
pub fn foo2() -> Result<Reexported, ()> { Ok(0) }
```
